### PR TITLE
feature/#6766 JSON and Startree index information in API

### PR DESCRIPTION
**Pull request by kriti-sc**
This PR exposes information about JSON index and star tree index in the segment metadata API.

Sample response:
```
"columns": [],
"indexes": {
    "studentID": {
        "bloom-filter": "NO",
        "dictionary": "YES",
        "forward-index": "YES",
        "inverted-index": "YES",
        "null-value-vector-reader": "NO",
        "range-index": "NO",
        "json-index": "NO"
    },
    "firstName": {
        "bloom-filter": "NO",
        "dictionary": "YES",
        "forward-index": "YES",
        "inverted-index": "NO",
        "null-value-vector-reader": "NO",
        "range-index": "NO",
        "json-index": "NO"
    }
},
"startree-index": [
    {
        "dimension-columns": [
            "studentID",
            "firstName"
        ],
        "metric-aggregations": [
            {
                "column": "col1",
                "aggregation": "SUM"
            },
            {
                "column": "col2",
                "aggregation": "DISTINCT"
            },
            {
                "column": "*",
                "aggregation": "COUNT"
            }
        ],
        "max-leaf-records": 1,
        "dimension-columns-skipped": [
            "col3"
        ]
    }
]
```

### CODE REVIEW

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java [hunk 1]
```diff
@@ -63,9 +77,67 @@ public static String getSegmentMetadata(SegmentDataManager segmentDataManager, L
...
     return JsonUtils.objectToString(segmentMetadataJson);
   }
 
+  /**
+   * Get the JSON object containing star tree index details for a segment.
+   */
+  private static List<Map<String, Object>> getStartreeIndexForSegmentColumns(SegmentDataManager segmentDataManager) {
+    List<Map<String, Object>> startreeDetails = new ArrayList<>();
+
+    if (segmentDataManager instanceof ImmutableSegmentDataManager) {
```
Comment (Jackie-Jiang) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:90]: We don't need these `instanceof` checks. The method is declared in the interface

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java [hunk 2]
```diff
@@ -63,9 +77,67 @@ public static String getSegmentMetadata(SegmentDataManager segmentDataManager, L
...
+      StarTreeV2Metadata starTreeMetadata = starTree.getMetadata();
+
+      Map<String, Object> starTreeIndexMap = new LinkedHashMap<>();
+
+      List<String> starTreeDimensions = starTreeMetadata.getDimensionsSplitOrder();
+      starTreeIndexMap.put(STAR_TREE_DIMENSION_COLUMNS, starTreeDimensions);
+
+      List<Map<String, String>> starTreeMetricAggregations = new ArrayList<>();
+      Set<AggregationFunctionColumnPair> functionColumnPairs = starTreeMetadata.getFunctionColumnPairs();
+      for (AggregationFunctionColumnPair functionColumnPair : functionColumnPairs) {
```
Comment (Jackie-Jiang) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:123]: Directly put it as a list of `functionColumnPair.toColumnName()`, no need to create a separate map

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java (hunk 1 above)
Comment (kriti-sc) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java]: This is done.

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java (hunk 2 above)
Comment (kriti-sc) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:114]: This is done.

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java [hunk 3]
```diff
@@ -62,24 +76,63 @@ public static String getSegmentMetadata(SegmentDataManager segmentDataManager, L
...
-    segmentMetadataJson.set("indexes", JsonUtils.objectToJsonNode(getIndexesForSegmentColumns(segmentDataManager)));
+    segmentMetadataJson.set(COLUMN_INDEX_KEY, JsonUtils.objectToJsonNode(getIndexesForSegmentColumns(segmentDataManager)));
+    segmentMetadataJson.set(STAR_TREE_INDEX_KEY, JsonUtils.objectToJsonNode((getStartreeIndexForSegmentColumns(segmentDataManager))));
     return JsonUtils.objectToString(segmentMetadataJson);
   }
 
+  /**
+   * Get the JSON object containing star tree index details for a segment.
+   */
+  private static List<Map<String, Object>> getStartreeIndexForSegmentColumns(SegmentDataManager segmentDataManager) {
```
Comment (Jackie-Jiang) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:87]: (nit) Suggest putting the star-tree related methods after the indexes related methods

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java [hunk 4]
```diff
@@ -62,24 +76,63 @@ public static String getSegmentMetadata(SegmentDataManager segmentDataManager, L
...
+    List<Map<String, Object>> startreeDetails = new ArrayList<>();
+    ImmutableSegmentDataManager immutableSegmentDataManager = (ImmutableSegmentDataManager) segmentDataManager;
+    ImmutableSegment immutableSegment = immutableSegmentDataManager.getSegment();
+    List<StarTreeV2> starTrees = immutableSegment.getStarTrees();
+    if (starTrees == null) {
+      return startreeDetails;
+    }
+    startreeDetails = getImmutableSegmentStartreeIndexes(starTrees);
+    return startreeDetails;
```
Comment (Jackie-Jiang) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:96]: Can be simplified. Also suggest returning `null` if index does not exist
```suggestion
    List<StarTreeV2> starTrees = segmentDataManager.getSegment().getStarTrees();
    return starTrees != null ? getImmutableSegmentStartreeIndexes(starTrees) : null;
```

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java [hunk 5]
```diff
@@ -62,24 +76,63 @@ public static String getSegmentMetadata(SegmentDataManager segmentDataManager, L
...
-        ImmutableSegmentImpl immutableSegmentImpl = (ImmutableSegmentImpl) immutableSegment;
-        Map<String, ColumnIndexContainer> columnIndexContainerMap = immutableSegmentImpl.getIndexContainerMap();
-        columnIndexMap = getImmutableSegmentColumnIndexes(columnIndexContainerMap);
-      }
-    }
+    ImmutableSegmentDataManager immutableSegmentDataManager = (ImmutableSegmentDataManager) segmentDataManager;
+    ImmutableSegment immutableSegment = immutableSegmentDataManager.getSegment();
+    ImmutableSegmentImpl immutableSegmentImpl = (ImmutableSegmentImpl) immutableSegment;
+    Map<String, ColumnIndexContainer> columnIndexContainerMap = immutableSegmentImpl.getIndexContainerMap();
+    columnIndexMap = getImmutableSegmentColumnIndexes(columnIndexContainerMap);
```
Comment (Jackie-Jiang) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:135]: Need to keep the check before casting here
```suggestion
    IndexSegment segment = segmentDataManager.getSegment();
    if (segment instanceof ImmutableSegmentImpl) {
      return getImmutableSegmentColumnIndexes(((ImmutableSegmentImpl) segment).getIndexContainerMap());
    } else {
      return null;
    }
```

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java (hunk 3 above)
Comment (kriti-sc) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:87]: This is done.

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java (hunk 4 above)
Comment (kriti-sc) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:96]: This is done

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java [hunk 6]
```diff
@@ -128,8 +137,49 @@ public static String getSegmentMetadata(SegmentDataManager segmentDataManager, L
...
+
       columnIndexMap.put(entry.getKey(), indexStatus);
     }
     return columnIndexMap;
   }
+
+  /**
+   * Get the JSON object containing star tree index details for a segment.
+   */
+  private static List<Map<String, Object>> getStartreeIndexForSegmentColumns(SegmentDataManager segmentDataManager) {
```
Comment (Jackie-Jiang) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:154]:
```suggestion
  @Nullable
  private static List<Map<String, Object>> getStarTreeIndexes(SegmentDataManager segmentDataManager) {
```

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java [hunk 7]
```diff
@@ -62,25 +75,21 @@ public static String getSegmentMetadata(SegmentDataManager segmentDataManager, L
...
-    segmentMetadataJson.set("indexes", JsonUtils.objectToJsonNode(getIndexesForSegmentColumns(segmentDataManager)));
+    segmentMetadataJson.set(COLUMN_INDEX_KEY, JsonUtils.objectToJsonNode(getIndexesForSegmentColumns(segmentDataManager)));
+    segmentMetadataJson.set(STAR_TREE_INDEX_KEY, JsonUtils.objectToJsonNode((getStartreeIndexForSegmentColumns(segmentDataManager))));
     return JsonUtils.objectToString(segmentMetadataJson);
   }
 
   /**
    * Get the JSON object with the segment column's indexing metadata.
    */
   private static Map<String, Map<String, String>> getIndexesForSegmentColumns(SegmentDataManager segmentDataManager) {
```
Comment (Jackie-Jiang) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:86]: Annotate with nullable

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java [hunk 8]
```diff
@@ -128,8 +137,49 @@ public static String getSegmentMetadata(SegmentDataManager segmentDataManager, L
...
+   */
+  private static List<Map<String, Object>> getStartreeIndexForSegmentColumns(SegmentDataManager segmentDataManager) {
+    List<StarTreeV2> starTrees = segmentDataManager.getSegment().getStarTrees();
+    return starTrees != null ? getImmutableSegmentStartreeIndexes(starTrees) : null;
+  }
+
+  /**
+   * Helper to loop over star trees of a segment to create a map containing star tree details.
+   */
+  private static List<Map<String, Object>> getImmutableSegmentStartreeIndexes(List<StarTreeV2> starTrees){
```
Comment (Jackie-Jiang) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:162]:
```suggestion
  private static List<Map<String, Object>> getStarTreeIndexes(List<StarTreeV2> starTrees){
```

File: pinot-server/src/test/java/org/apache/pinot/server/api/TablesResourceTest.java [hunk 9]
```diff
@@ -273,6 +273,7 @@ public void testOfflineTableSegmentMetadata()
             .get(String.class));
     Assert.assertEquals(jsonResponse.get("columns").size(), 2);
     Assert.assertEquals(jsonResponse.get("indexes").size(), 17);
+    Assert.assertEquals(jsonResponse.get("star-tree-index").size(), 0);
```
Comment (Jackie-Jiang) [inline pinot-server/src/test/java/org/apache/pinot/server/api/TablesResourceTest.java:276]: You might need to modify the test because it can be `null` here

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java (hunk 7 above)
Comment (kriti-sc) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:86]: This is done

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java (hunk 8 above)
Comment (kriti-sc) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:163]: This is done

Comment (kriti-sc) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java]: This is resolved in a different way than suggested, please do take a look. @Jackie-Jiang

File: pinot-server/src/test/java/org/apache/pinot/server/api/TablesResourceTest.java (hunk 9 above)
Comment (kriti-sc) [reply pinot-server/src/test/java/org/apache/pinot/server/api/TablesResourceTest.java]: Fixed this by using `Assert.assertNull`

**Comment by npawar** [comment]
Hey @kriti-sc please resolve the merge conflict and fix the test, then i'll merge this

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java [hunk 10]
```diff
@@ -128,8 +139,50 @@ public static String getSegmentMetadata(SegmentDataManager segmentDataManager, L
...
+  @Nullable
+  private static List<Map<String, Object>> getStarTreeIndexesForSegment(SegmentDataManager segmentDataManager) {
+    List<StarTreeV2> starTrees = segmentDataManager.getSegment().getStarTrees();
+    return starTrees != null ? getStartreeIndexes(starTrees) : null;
+  }
+
+  /**
+   * Helper to loop over star trees of a segment to create a map containing star tree details.
+   */
+  private static List<Map<String, Object>> getStartreeIndexes(List<StarTreeV2> starTrees){
```
Comment (Jackie-Jiang) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java:165]: (nit)
```suggestion
  private static List<Map<String, Object>> getStarTreeIndexes(List<StarTreeV2> starTrees){
```

**Review by Jackie-Jiang — COMMENTED** [review]
Please take a look at the test failure

File: pinot-server/src/test/java/org/apache/pinot/server/api/TablesResourceTest.java (hunk 9 above)
Comment (kriti-sc) [reply pinot-server/src/test/java/org/apache/pinot/server/api/TablesResourceTest.java:276]: Reverted this.

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java (hunk 10 above)
Comment (kriti-sc) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/SegmentMetadataFetcher.java]: Fixed this.

**Comment by kriti-sc** [comment]
@Jackie-Jiang the tests are passing, this PR is ready to be merged.
